File: svc/nullblock-engrams/src/models/engram.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum EngramType {
    Persona,
    Preference,
    Strategy,
    Knowledge,
    Compliance,
}

impl EngramType {
    pub fn as_str(&self) -> &'static str {
        match self {
            EngramType::Persona => "persona",
            EngramType::Preference => "preference",
            EngramType::Strategy => "strategy",
            EngramType::Knowledge => "knowledge",
            EngramType::Compliance => "compliance",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "persona" => Some(EngramType::Persona),
            "preference" => Some(EngramType::Preference),
            "strategy" => Some(EngramType::Strategy),
            "knowledge" => Some(EngramType::Knowledge),
            "compliance" => Some(EngramType::Compliance),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Engram {
    pub id: Uuid,
    pub wallet_address: String,
    pub engram_type: String,
    pub key: String,
    pub tags: Vec<String>,
    pub content: serde_json::Value,
    pub summary: Option<String>,
    pub version: i32,
    pub parent_id: Option<Uuid>,
    pub lineage_root_id: Option<Uuid>,
    pub is_public: bool,
    pub is_mintable: bool,
    pub nft_token_id: Option<String>,
    pub price_mon: Option<rust_decimal::Decimal>,
    pub royalty_percent: Option<i32>,
    pub priority: i32,
    pub ttl_seconds: Option<i32>,
    pub created_by: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub accessed_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateEngramRequest {
    pub wallet_address: String,
    pub engram_type: String,
    pub key: String,
    pub content: serde_json::Value,
    #[serde(default)]
    pub tags: Vec<String>,
    pub summary: Option<String>,
    #[serde(default)]
    pub priority: i32,
    pub ttl_seconds: Option<i32>,
    pub created_by: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateEngramRequest {
    pub content: serde_json::Value,
    pub summary: Option<String>,
    pub reason: Option<String>,
    pub changed_by: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchEngramsRequest {
    pub wallet_address: Option<String>,
    pub engram_type: Option<String>,
    pub tags: Option<Vec<String>>,
    pub is_public: Option<bool>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForkEngramRequest {
    pub target_wallet: String,
    pub new_key: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct EngramHistory {
    pub id: Uuid,
    pub engram_id: Uuid,
    pub version: i32,
    pub content: serde_json::Value,
    pub changed_by: Option<String>,
    pub change_reason: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EngramResponse {
    pub success: bool,
    pub data: Option<Engram>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EngramsListResponse {
    pub success: bool,
    pub data: Vec<Engram>,
    pub total: i64,
    pub limit: i64,
    pub offset: i64,
}

File: svc/hecate/src/types/engram.ts
export type EngramType =
  | 'persona'
  | 'preference'
  | 'strategy'
  | 'knowledge'
  | 'compliance';

export interface Engram {
  id: string;
  wallet_address: string;
  engram_type: EngramType;
  key: string;
  content: string;
  metadata?: Record<string, unknown>;
  tags: string[];
  is_public: boolean;
  created_at: string;
  updated_at: string;
  summary?: string;
  version?: number;
  parent_id?: string;
  lineage_root_id?: string;
  is_mintable?: boolean;
  nft_token_id?: string;
  price_mon?: string;
  royalty_percent?: number;
  priority?: number;
  ttl_seconds?: number;
  created_by?: string;
  accessed_at?: string;
}

export interface EngramListResponse {
  engrams: Engram[];
  total: number;
}

export interface EngramSearchRequest {
  wallet_address?: string;
  engram_type?: EngramType;
  query?: string;
  tags?: string[];
  limit?: number;
  offset?: number;
}

export interface CreateEngramRequest {
  wallet_address: string;
  engram_type: EngramType;
  key: string;
  content: string;
  metadata?: Record<string, unknown>;
  tags?: string[];
  is_public?: boolean;
}

export type TransactionAction = 'buy' | 'sell';

export interface TransactionMetadata {
  graduation_progress?: number;
  holder_count?: number;
  volume_24h_sol?: number;
  market_cap_sol?: number;
  bonding_curve_percent?: number;
}

export interface TransactionSummary {
  tx_signature: string;
  action: TransactionAction;
  token_mint: string;
  token_symbol?: string;
  venue: string;
  entry_sol: number;
  exit_sol?: number;
  pnl_sol?: number;
  pnl_percent?: number;
  slippage_bps: number;
  execution_time_ms: number;
  strategy_id?: string;
  timestamp: string;
  metadata: TransactionMetadata;
}

export type ExecutionErrorType =
  | 'rpc_timeout'
  | 'slippage_exceeded'
  | 'insufficient_funds'
  | 'tx_failed'
  | 'simulation_failed'
  | 'signing_failed'
  | 'network_error'
  | 'invalid_params'
  | 'rate_limited'
  | 'unknown';

export interface ErrorContext {
  action?: string;
  token_mint?: string;
  attempted_amount_sol?: number;
  venue?: string;
  strategy_id?: string;
  edge_id?: string;
}

export interface ExecutionError {
  error_type: ExecutionErrorType;
  message: string;
  context: ErrorContext;
  stack_trace?: string;
  recoverable: boolean;
  timestamp: string;
}

export interface TradeHighlight {
  token: string;
  pnl_sol: number;
  tx_signature?: string;
}

export interface VenueMetrics {
  trades: number;
  pnl_sol: number;
  win_rate: number;
}

export interface StrategyMetrics {
  trades: number;
  pnl_sol: number;
  win_rate: number;
}

export interface DailyMetrics {
  period: string;
  total_trades: number;
  winning_trades: number;
  win_rate: number;
  total_pnl_sol: number;
  avg_trade_pnl: number;
  max_drawdown_percent: number;
  best_trade?: TradeHighlight;
  worst_trade?: TradeHighlight;
  by_venue: Record<string, VenueMetrics>;
  by_strategy: Record<string, StrategyMetrics>;
}

export const A2A_TAG_LEARNING = 'arbFarm.learning';

export interface EngramBrowserFilter {
  engram_type?: EngramType;
  tags?: string[];
  query?: string;
  limit?: number;
}
